### PR TITLE
Provide a simple, single-threaded abstraction of select, poll, epoll, kqueue, /dev/poll, event ports, and pollset

### PR DESCRIPTION
**Pull request by mattrm456**
This PR provides an abstract class, `ntsi::Reactor`, that is similar in purpose to `ntci::Reactor` but strips down the abstraction so that it only provides single-threaded registration of event interest and polling of events, without any callbacks. 

The purpose of such an abstraction is threefold:

1) Allow tests to be more-easily written that verifies of behavior of the various operating system APIs when sockets are in different conditions
2) Provide a low-level layer from which a new experimental implementation of NTC dynamic load balancing may be based
3) Make available a simple event loop abstraction for users that wish to only use NTS non-blocking sockets

### CODE REVIEW

File: groups/nts/ntsa/ntsa_event.h [hunk 1]
```diff
@@ -356,31 +696,37 @@ const ntsa::Error& Event::error() const
...
 bool Event::isWritable() const
 {
-    return ((d_state & (1 << e_WRITABLE)) != 0);
+    return ((d_state & (1U << e_WRITABLE)) != 0);
+}
+
+NTSCFG_INLINE
+bool Event::isExceptional() const
+{
+    return ((d_state & (1U << e_EXCEPTIONAL)) != 0);
```
Comment (smtrfnv) [inline groups/nts/ntsa/ntsa_event.h:711]: It is said that 
```
        /// The socket has an error or an exceptional condition.
        e_ERROR = 4,
```

Doesn't it mean that here it is needed also to check `e_ERROR` bit?

File: groups/nts/ntsa/ntsa_event.t.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+//                                 --------
+//
+//-----------------------------------------------------------------------------
+
+// [ 1]
+//-----------------------------------------------------------------------------
+// [ 1]
+//-----------------------------------------------------------------------------
+
+NTSCFG_TEST_CASE(1)
```
Comment (smtrfnv) [inline groups/nts/ntsa/ntsa_event.t.cpp:34]: It looks like that this test is sort of unfinished.

File: groups/nts/ntsa/ntsa_interest.h [hunk 3]
```diff
@@ -0,0 +1,1189 @@
...
+
+    /// Define a type alias for of vector of interest, indexed by socket
+    /// handle.
+    typedef bsl::vector<ntsa::Interest> Vector;
+
+    /// Define a type alias for a set of socket handles.
+    typedef bsl::set<ntsa::Handle> Set;
+
+    Vector            d_vector;
+    Set               d_set;
```
Comment (smtrfnv) [inline groups/nts/ntsa/ntsa_interest.h:199]: As far as I got it `d_set` is used to only prevent a socket to be attached twice, to control detachment process and to count numer of attached sockets. Probably, the same thing can be achieved if type of `Vector` is changed to something like `typedef bsl::vector<bsl::pair<bool,. ntsa::Interest> > Vector;` where the first member of the pair means status of the socket (attached/detached) and a separate socket counter.

File: groups/nts/ntsa/ntsa_interest.t.cpp [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+//                                 --------
+//
+//-----------------------------------------------------------------------------
+
+// [ 1]
+//-----------------------------------------------------------------------------
+// [ 1]
+//-----------------------------------------------------------------------------
+
+NTSCFG_TEST_CASE(1)
```
Comment (smtrfnv) [inline groups/nts/ntsa/ntsa_interest.t.cpp:34]: Is this test finished? maybe we can add some validation?

File: groups/nts/ntsu/ntsu_socketutil.cpp [hunk 5]
```diff
@@ -7788,6 +7788,23 @@ bsl::size_t SocketUtil::maxBacklog()
 
 bool SocketUtil::isSocket(ntsa::Handle socket)
```
Comment (smtrfnv) [inline groups/nts/ntsu/ntsu_socketutil.cpp:7789]: By the way, I did not find any test for this function. I think it's relatively easy to test it, so let's do it.

File: groups/nts/ntso/ntso_select.cpp [hunk 6]
```diff
@@ -0,0 +1,639 @@
...
+Select::~Select()
+{
+    NTSO_SELECT_LOG_DEVICE_CLOSE();
+}
+
+ntsa::Error Select::attachSocket(ntsa::Handle socket)
+{
+    ntsa::Error error;
+
+    if (socket < 0) {
```
Comment (smtrfnv) [inline groups/nts/ntso/ntso_select.cpp:269]: The same validation is done inside `d_interestSet.attach(socket);`

File: groups/nts/ntso/ntso_select.cpp [hunk 7]
```diff
@@ -0,0 +1,639 @@
...
+    NTSO_SELECT_LOG_ADD(socket);
+
+    return ntsa::Error();
+}
+
+ntsa::Error Select::detachSocket(ntsa::Handle socket)
+{
+    ntsa::Error error;
+
+    if (socket < 0) {
```
Comment (smtrfnv) [inline groups/nts/ntso/ntso_select.cpp:293]: The same validation is done inside `d_interestSet.detach(socket);`

File: groups/nts/ntsa/ntsa_event.h (hunk 1 above)
Comment (mattrm456) [reply groups/nts/ntsa/ntsa_event.h]: An e_ERROR does not imply e_EXCEPTIONAL, this is mistake made when revising documentation from NTC. I have fixed this.

File: groups/nts/ntsa/ntsa_event.t.cpp (hunk 2 above)
Comment (mattrm456) [reply groups/nts/ntsa/ntsa_event.t.cpp]: I have committed an expanded test driver.

File: groups/nts/ntsa/ntsa_interest.h (hunk 3 above)
Comment (mattrm456) [reply groups/nts/ntsa/ntsa_interest.h:252]: ntsa::Interest and ntsa::Event and their respective sets are designed to store and merge what we want (ntsa::Interest) with what we have (ntsa::Event) with some guarantees of order of traversal. I want to position these components to be usable in both test drivers, where clarity of operating system behavior is paramount, as well as aiding us in revising our implementation of dynamic load balancing in NTC, where performance is critical. Right now, the interest and event sets are implemented very simply, to provide the basic design. Next up is a revision of their implementations into something like a giant bit set, leveraging the fact that file descriptor numbers are always kept fairly small, while still providing us O(1) lookups and guaranteed order of traversal (and perhaps queue-like behavior ala epoll and kqueue, but that may be getting ahead of ourselves.)

File: groups/nts/ntsa/ntsa_interest.t.cpp (hunk 4 above)
Comment (mattrm456) [reply groups/nts/ntsa/ntsa_interest.t.cpp:138]: I have committed an expanded test driver.

File: groups/nts/ntsu/ntsu_socketutil.cpp (hunk 5 above)
Comment (mattrm456) [reply groups/nts/ntsu/ntsu_socketutil.cpp:7789]: I have added a test for ntsu::SocketUtil::isSocket.

File: groups/nts/ntso/ntso_select.cpp (hunk 6 above)
Comment (mattrm456) [reply groups/nts/ntso/ntso_select.cpp]: We now defend against invalid handles only in ntsa::InterestSet.

File: groups/nts/ntso/ntso_select.cpp (hunk 7 above)
Comment (mattrm456) [reply groups/nts/ntso/ntso_select.cpp]: We now defend against invalid handles only in ntsa::InterestSet.

**Comment by mattrm456** [comment]
I have addressed all comments. Please re-review.

**Review by smtrfnv — APPROVED** [review]
LGTM. I like the simplicity of the reactors

**Pull request merged**